### PR TITLE
Fix race conditions in thread management.

### PR DESCRIPTION
**Pull request by mpoeter**
The crash dump analysis for #7833 showed a race condition in the destructor of `ApplierThread`. The immediate workaround was simple enough (move the `reset` call after `shutdown`), but the underlying problem still remained: it is possible that (parts) of a thread's destructor are executed while the thread is still running.

While digging through the code to understand how the thread management currently works I found a few other cases that looked a lot like subtle race conditions to me. This PR is an attempt to fix all this issues that I've found.

The most important change is that `shutdown` now blocks until the thread has terminated. It is important that the _most derived_ class calls shutdown in the destructor. Usually this should be the very first thing since accesses to class members that happen before that call would have to be synchronized with the thread itself. The reason why `shutdown` must be the called by the most derived class is that otherwise we would have a race on the class's vtable.

Another small race was in `Thread::start` where `_threadStructInitialized` is set _after_ starting the thread. It could happen that the thread terminates before this member is set. If `_deleteOnExit` is set this would either result in a resource leak (best case) or a segfault. This is solved by using an explicit reference counter, so the thread only calls delete on itself if there are no more pending references.

### CODE REVIEW

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2843/

File: lib/Basics/Thread.cpp [hunk 1]
```diff
@@ -182,62 +178,37 @@ void Thread::beginShutdown() {
     _state.compare_exchange_strong(state, ThreadState::STOPPED);
   }
 
-  while (state != ThreadState::STOPPING && state != ThreadState::STOPPED &&
-         state != ThreadState::DETACHED) {
+  while (state != ThreadState::STOPPING && state != ThreadState::STOPPED) {
     _state.compare_exchange_strong(state, ThreadState::STOPPING);
```
Comment (ObiWahn) [inline lib/Basics/Thread.cpp]: Could we use the weak variant in the loop?

Comment (mpoeter) [reply lib/Basics/Thread.cpp]: Yes, since this is inside a loop the weak variant would be sufficient.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2849/

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2859/

**Comment by mpoeter** [comment]
@jsteemann thanks! I've already made the same changes in my fork :) https://github.com/mpoeter/arangodb/commit/d29ebab4557759836377d9139307013493da95f0
Unfortunately these commits do not show up in the PR unless someone from your team explicitly pulls them.

**Comment by jsteemann** [comment]
@mpoeter : ok, thanks. Is there anything else I should pull from your fork, or was it only about the strong-to-weak changes?

**Comment by jsteemann** [comment]
@mpoeter : ok, thanks. I merged your changes into the arangodb bugfix branch now.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2862/

**Pull request merged**